### PR TITLE
Add support for missing return obj from to function on custom batch objects

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds support for moving custom data object on device where the object doesn't return self from the to function.

Example: https://github.com/intel-isl/Open3D-ML/blob/master/ml3d/torch/dataloaders/concat_batcher.py#L484

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/apply_func.py [hunk 1]
```diff
@@ -245,7 +245,11 @@ def batch_to(data):
             return device_data
 
         kwargs = dict(non_blocking=True) if isinstance(data, torch.Tensor) else {}
-        return data.to(device, **kwargs)
+        data_output = data.to(device, **kwargs)
+        if data_output is not None:
+            return data_output
+        # user wrongly implemented the ``TransferableDataType`` and forgot to return ``self``.
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:251]: nit: double backticks are for rendered docs

```suggestion
        # user wrongly implemented the `TransferableDataType` and forgot to return `self`.
```

**Review by carmocca — APPROVED** [review]
Not a fan of us doing this but I guess it doesn't hurt. A tool like mypy would have reported the bug

File: tests/utilities/test_apply_func.py [hunk 2]
```diff
@@ -209,3 +209,22 @@ def fn(a, b):
...
+    class TensorObject:
+
+        def __init__(self, tensor: torch.Tensor, should_return: bool = True):
+            self.tensor = tensor
+            self.should_return = should_return
+
+        def to(self, device):
+            self.tensor.to(device)
+            if self.should_return:
+                return self
```
Comment (awaelchli) [inline tests/utilities/test_apply_func.py:226]:
```suggestion
            # simulate a user forgets to return self
            if self.should_return:
                return self
```

**Review by awaelchli — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by justusschock — APPROVED** [review]
